### ISSUE TITLE
systemd-networkd timing problem when adding routes to a dhcp interface

### ISSUE THREAD

**Issue by dheule** [systemd/systemd#1850]
I have a setup where i need to get the ip via dhcp, but we must add some static routes manual

``` ini
[Match]
Name=bridge0

[Network]
DHCP=ipv4

[Route]
Gateway=192.168.0.1
Destination=10.0.0.0/8

[Route]
Gateway=192.168.0.1
Destination=192.168.100.0/24
```

It works when systemd-networkd is restarted.

The problem seams to be that the routes are inserted bevor the dhcp client has a ip set to the interface:

``` log
Nov 11 14:06:05 linuxhost systemd-networkd[341]: bridge0: netdev ready
Nov 11 14:06:05 linuxhost systemd-networkd[341]: Enumeration completed
Nov 11 14:06:05 linuxhost systemd-networkd[341]: bridge0: bridge0         : could not set route: Network is unreachable
Nov 11 14:06:05 linuxhost systemd-networkd[341]: bridge0: bridge0         : could not set route: Network is unreachable
Nov 11 14:06:05 linuxhost systemd-networkd[341]: enp13s0: Configured
Nov 11 14:06:08 linuxhost systemd-networkd[341]: enp13s0: Gained carrier
Nov 11 14:06:08 linuxhost systemd-networkd[341]: bridge0: Gained carrier
Nov 11 14:06:11 linuxhost systemd-networkd[341]: bridge0: DHCPv4 address 192.168.0.22/24 via 192.168.0.1
Nov 11 14:06:11 linuxhost systemd-networkd[341]: bridge0: Configured
```

The order should be to first configure the interface via dhcp and then add the routes ...

Thank you
Daniel

**Comment by fredrikwidlund** [comment]
I have a similar issue. 

I am using a 4x10GbE 802.3ad bond as a trunk with a VLAN on top.

A VLAN configuration looks like below:

```
[Match]
Name=bond1.100

[Address]
Address=1.2.3.4/24

[Route]
Destination=0.0.0.0/0
Gateway=1.2.3.1
```

I also have 10/8 routes on other VLANs etc.

Everything comes up all right (except for the routes) but it takes some seconds for the bonding interface to come alive, and the default route, as well as other routes, are not added. After systemctl restart everything comes up find.

I don't see anything obviously related in the journal.

Kind regards,
Fredrik

**Comment by f0** [comment]
@poettering  any news here? 
We also see this problem with a bonding interface, its exactly the same as @fredrikwidlund  describes
(we have no dhcp running)

**Comment by kozmagabor** [comment]
I had the same problem too @f0 , but i found a solution and it's works on 16.04 ubuntu.
It's lacp bondig config with vlan without dhcp.
4 port 10Gbe

```
# ~: tail -n 100 -f /etc/systemd/network/*
==> /etc/systemd/network/bond0-eth0.network <==
[Match]
Name=eth0

[Network]
Bond=bond0
DHCP=no
LinkLocalAddressing=no

==> /etc/systemd/network/bond0-eth1.network <==
[Match]
Name=eth1

[Network]
Bond=bond0
DHCP=no
LinkLocalAddressing=no

==> /etc/systemd/network/bond0-eth2.network <==
[Match]
Name=eth2

[Network]
Bond=bond0
DHCP=no
LinkLocalAddressing=no

==> /etc/systemd/network/bond0-eth3.network <==
[Match]
Name=eth3

[Network]
Bond=bond0
DHCP=no
LinkLocalAddressing=no

==> /etc/systemd/network/bond0.netdev <==
[NetDev]
Name=bond0
Kind=bond

[Bond]
Mode=802.3ad
TransmitHashPolicy=layer3+4
MIIMonitorSec=1s
LACPTransmitRate=fast
UpDelaySec=2s
DownDelaySec=4s

==> /etc/systemd/network/bond0.network <==
[Match]
Name=bond0

[Network]
VLAN=vlan101
DHCP=no
LinkLocalAddressing=false
IPv4LLRoute=true
LLMNR=false
==> /etc/systemd/network/bond0-vlan101.netdev <==
[NetDev]
Name=vlan101
Kind=vlan

[VLAN]
Id=101

==> /etc/systemd/network/bond0-vlan101.network <==
[Match]
Name=vlan101

[Network]
DHCP=no
Address=xxx.xxx.xxx.xxx/24
Gateway=xxx.xxx.xxx.xxx
```

Because the systemd-networkd doesn't ovverride default bonding settings on bond0

```
==> /etc/modprobe.d/bonding.conf <==
options bonding max_bonds=0
```

**Comment by f0** [comment]
@kozmagabor  thx for the hints, sadly they do not work for us..
Tested with CoreOS 1010.1.0 (systemd 225) and CoreOS 1032.0.0 (systemd 229)

It all works as excepted after restarting systemd-networkd... but on the next boot its broken again

**Comment by f0** [comment]
any chance to get this fixed? (i am not a c guy, so i can not fix it myself)

**Comment by vimagick** [comment]
I had the same problem too. Static routes was flushed by dhcp.

**Comment by ssahani** [comment]
1. Via the DHCP it takes a while and we also sets the static routes asynchronously.
2. So If I figure this this requirement right the DHCP routes are set then the static routes supposed to be configured. I think we need to introduce a config `ConfigureStaticRoutesAfterDHCP`. When this is set static routes will be configured after the DHCP routes are set.

**Comment by amessina** [comment]
@ssahani, does a new config need to exist?  If static routes are configured, shouldn't that be enough to say that the admin wants them there?

**Comment by flo-wer** [comment]
This should work with GatewayOnlink=yes in [Route], see https://www.freedesktop.org/software/systemd/man/systemd.network.html#GatewayOnlink=
The problem is whenever a new route is inserted, the kernel checks if the gateway is reachable. If you insert a route while no interface is configured this fails:
```
# ip route add fd01::/64 via fd00::1 
RTNETLINK answers: No route to host
# ip route add 192.168.100.0/24 via 192.168.0.1 
RTNETLINK answers: Network is unreachable
```
This is the case when using DHCP, as the default route is not available at the time, the static route is created.

This was already reported in #1283 and fixed with #5734.

**Comment by ssahani** [comment]
@flo-wer thanks . That should work. No need for anything else.

**Comment by aritradalal** [comment]
This also occurs under the following circumstance:
```
[Match]
Name=remoter

[Network]
DHCP=no
Address=10.255.253.4/32

[Route]
Destination=10.255.253.1
Scope=link

[Route]
Destination=10.255.254.0/23
Gateway=10.255.253.1
Scope=global
GatewayOnlink=yes
```
However, the GatewayOnlink does not seem to be adding anything to the route list. Instead, I see this in the logs:

`Oct 22 23:09:30 server0.hidden-domain.com systemd-networkd[21078]: remoter: Could not set route: Network is unreachable`

**Comment by morphinz** [comment]
Same here.

Jan 29 16:06:02 AA-FKM2 systemd-networkd[26053]: data: Could not set route: Network is unreachable
Jan 29 16:06:02 AA-FKM2 systemd-networkd[26053]: data: Could not set route: Network is unreachable
Jan 29 16:06:02 AA-FKM2 systemd-networkd[26053]: data: Could not set route: Network is unreachable

**Comment by Chupaka** [comment]
So, still no fix or at least workaround? :(

**Comment by ftoral** [comment]
I came along with the same issue when managing network configured by Netplan.
This bug is cross referenced here: https://bugs.launchpad.net/netplan/+bug/1781652

**Comment by wsw70** [comment]
> This should work with GatewayOnlink=yes in [Route], see https://www.freedesktop.org/software/systemd/man/systemd.network.html#GatewayOnlink=

@flo-wer thank you for the solution / workaround

> This was already reported in #1283 and fixed with #5734.

Does it mean that the current behaviour (failing to set a route when the link is not there yet) should be seen as correct? (I am not sure if  `GatewayOnlink` is a solution / extension of the otherwise proper behaviour of `systemd-networkd`, or a workaround)

**Comment by ssahani** [comment]
Is is reproducible with current git ?

**Comment by yuwata** [comment]
This should be fixed by using `GatewayOnlink=`, as mentioned above. Let's close this. If the option does not work, then please provide debug logs of networkd and relevant configs. Thank you.

**Comment by myrdd** [comment]
some notes (FYI)
* `GatewayOnlink` is available since v234
* the option has been renamed to `GatewayOnLink` in v242-rc1. (https://github.com/systemd/systemd/commit/9cb8c5593443d24c19e40bfd4fc06d672f8c554c)

**Comment by letoams** [comment]
I have another variant of this problem, where for some reason eth0 boots slower than eth1 (both are configured staticly):
Jun 19 17:21:19 nic systemd-networkd[325]: eth1: Gained carrier
Jun 19 17:21:19 nic systemd-networkd[325]: eth1: Could not set route: No route to host
Jun 19 17:21:19 nic systemd-networkd[325]: eth1: Failed
Jun 19 17:21:19 nic systemd-networkd[325]: eth0: IPv6 successfully enabled
Jun 19 17:21:19 nic systemd-networkd[325]: eth0: Link UP
Jun 19 17:21:19 nic systemd-networkd[325]: eth2: Link UP

because eth0 is not up when eth1 is doing routes, the entire eth1 interface fails.

For some reason this started happening only after running a dnf update on my existing fedora32 today

**Comment by wzrdtales** [comment]
same here, `PreferredSource` will cause the same problem.

**Comment by yuwata** [comment]
@letoams and @wzrdtales, could you provide the debugging logs of the networkd and open a new issue? Otherwise, we can easily miss such already closed and old issue page... You can generate debugging logs by creating the following drop-in config:
```
# /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
